src/HID-APIs/DefaultKeyboardAPI.hpp

```diff
@@ -42,26 +42,53 @@ size_t DefaultKeyboardAPI::set(KeyboardKeycode k, bool s)
 	}
 	// Its a normal key
 	else{
-		// Add k to the key report only if it's not already present
-		// and if there is an empty slot. Remove the first available key.
-		for (uint8_t i = 0; i < sizeof(_keyReport.keycodes); i++)
-		{
-			auto key = _keyReport.keycodes[i];
-			
-			// Is key already in the list or did we found an empty slot?
-			if (s && (key == uint8_t(k) || key == KEY_RESERVED)) {
-				_keyReport.keycodes[i] = k;
-				return 1;
+		// get size of keycodes during compile time
+		const uint8_t keycodesSize = sizeof(_keyReport.keycodes);
+
+		// if we are adding an element to keycodes
+		if (s){
+			// iterate through the keycodes
+			for (uint8_t i = 0; i < keycodesSize; i++)
+			{
+				auto key = _keyReport.keycodes[i];
+
+				// write to keycodes on first instance of k or null in keycodes
+				if (key == uint8_t(k) || key == KEY_RESERVED) {
+					_keyReport.keycodes[i] = k;
+					return 1;
+				}
 			}
-			
-			// Test the key report to see if k is present. Clear it if it exists.
-			if (!s && (key == k)) {
-				_keyReport.keycodes[i] = KEY_RESERVED;
-				return 1;
+		} else { // we are removing k from keycodes
+			// this will replace the first instance of k with the last entry in the pseudo-linkedlist
+			// if list is not empty
+			if (_keyReport.keycodes[0] != KEY_RESERVED) {
+				uint8_t keyIndex = 255; // if keyIndex == 255 then k does not exist in keycodes
+				KeyboardKeycode lastElement;
+				for (uint8_t i = 0; i < keycodesSize; i++)
+				{
+					auto key = _keyReport.keycodes[i];
+
+					if (key == uint8_t(k)) {
+						keyIndex = i;
+						continue;
+					}
+
+					// if the next element is null or if this is the last element
+					if ((i < keycodesSize-1 && _keyReport.keycodes[i+1] == KEY_RESERVED) || i == keycodesSize-1) {
```

Comment:
Thanks, my previous commit (e2145e8) did not have this bug, it got introduced in the "refactor" commit (b733010). 

As per your suggestion `lastIndex` is recorded and handled in the target key found block instead. I rewrote some tests and confirmed this bug is solved in the most recent commit. It also looks more readable so that is nice too.